qa/tasks/cephadm.py

```diff
@@ -123,83 +125,154 @@ def download_cephadm(ctx, config, ref):
-            )
-        elif git_url.startswith('https://github.com/'):
-            # git archive doesn't like https:// URLs, which we use with github.
-            rest = git_url.split('https://github.com/', 1)[1]
-            rest = re.sub(r'\.git/?$', '', rest).strip() # no .git suffix
-            ctx.cluster.run(
-                args=[
-                    'curl', '--silent',
-                    'https://raw.githubusercontent.com/' + rest + '/' + ref + '/src/cephadm/cephadm',
-                    run.Raw('>'),
-                    ctx.cephadm,
-                    run.Raw('&&'),
-                    'ls', '-l',
-                    ctx.cephadm,
-                ],
-            )
+            _fetch_cephadm_from_rpm(ctx)
+        # TODO: come up with a sensible way to detect if we need an "old, uncompiled"
+        # cephadm
+        elif 'cephadm_git_url' in config and 'cephadm_branch' in config:
+            _fetch_cephadm_from_github(ctx, config, ref)
         else:
-            ctx.cluster.run(
-                args=[
-                    'git', 'archive',
-                    '--remote=' + git_url,
-                    ref,
-                    'src/cephadm/cephadm',
-                    run.Raw('|'),
-                    'tar', '-xO', 'src/cephadm/cephadm',
-                    run.Raw('>'),
-                    ctx.cephadm,
-                ],
-            )
-        # sanity-check the resulting file and set executable bit
-        cephadm_file_size = '$(stat -c%s {})'.format(ctx.cephadm)
+            _fetch_cephadm_from_chachra(ctx, config, cluster_name)
+
+    try:
+        yield
+    finally:
+        _rm_cluster(ctx, cluster_name)
+        if config.get('cephadm_mode') == 'root':
+            _rm_cephadm(ctx)
+
+
+def _fetch_cephadm_from_rpm(ctx):
+    log.info("Copying cephadm installed from an RPM package")
+    # cephadm already installed from redhat.install task
+    ctx.cluster.run(
+        args=[
+            'cp',
+            run.Raw('$(which cephadm)'),
+            ctx.cephadm,
+            run.Raw('&&'),
+            'ls', '-l',
+            ctx.cephadm,
+        ]
+    )
+
+
+def _fetch_cephadm_from_github(ctx, config, ref):
+    ref = config.get('cephadm_branch', ref)
+    git_url = config.get('cephadm_git_url', teuth_config.get_ceph_git_url())
+    log.info('Downloading cephadm (repo %s ref %s)...' % (git_url, ref))
+    if git_url.startswith('https://github.com/'):
+        # git archive doesn't like https:// URLs, which we use with github.
+        rest = git_url.split('https://github.com/', 1)[1]
+        rest = re.sub(r'\.git/?$', '', rest).strip() # no .git suffix
         ctx.cluster.run(
             args=[
-                'test', '-s', ctx.cephadm,
-                run.Raw('&&'),
-                'test', run.Raw(cephadm_file_size), "-gt", run.Raw('1000'),
+                'curl', '--silent',
+                'https://raw.githubusercontent.com/' + rest + '/' + ref + '/src/cephadm/cephadm',
```

Comment:
Does this code really work? For `main` and `reef` I'm getting:

```
$ curl https://raw.githubusercontent.com/ceph/ceph/main/src/cephadm/cephadm
404: Not Found
```

Reply:
The `download_cephadm` function calls either _fetch_cephadm_from_github (what you highlighted) or _fetch_cephadm_from_chachra*. The former should generally be called when teuthololgy needs to check out an old version of cephadm. The latter should be called when a newer cephadm is needed.

There's a comment in `download_cephadm` that says `TODO: come up with a sensible way to detect if we need an "old, uncompiled" cephadm`. It is possible you are hitting a case where the detection is not right.  Maybe file a tracker issue and CC me (and Adam King) and I'll try to help some more when I return to work next week.



*I just noticed this is also a typo.

Reply:
Hi @phlogistonjohn!

The problem has popped up during the reef2squid upgrade testing: https://pulpito.ceph.com/yuriw-2023-11-23_15:43:47-upgrade:reef-x-wip-yuri10-testing-2023-11-22-1112-distro-default-smithi/.

All the jobs in the `upgrade:reef-x` suite failed due to:

```
023-11-23T16:06:23.673 DEBUG:teuthology.orchestra.run.smithi119:> sudo hostname $(hostname -s)
2023-11-23T16:06:23.700 INFO:tasks.cephadm:Downloading cephadm (repo https://github.com/ceph/ceph ref reef)...
2023-11-23T16:06:23.701 DEBUG:teuthology.orchestra.run.smithi114:> curl --silent https://raw.githubusercontent.com/ceph/ceph/reef/src/cephadm/cephadm > /home/ubuntu/cephtest/cephadm && ls -l /home/ubuntu/cephtest/cephadm
2023-11-23T16:06:23.923 INFO:teuthology.orchestra.run.smithi114.stdout:-rw-rw-r--. 1 ubuntu ubuntu 14 Nov 23 16:06 /home/ubuntu/cephtest/cephadm
2023-11-23T16:06:23.924 DEBUG:teuthology.orchestra.run.smithi119:> curl --silent https://raw.githubusercontent.com/ceph/ceph/reef/src/cephadm/cephadm > /home/ubuntu/cephtest/cephadm && ls -l /home/ubuntu/cephtest/cephadm
2023-11-23T16:06:24.018 INFO:teuthology.orchestra.run.smithi119.stdout:-rw-rw-r--. 1 ubuntu ubuntu 14 Nov 23 16:06 /home/ubuntu/cephtest/cephadm
2023-11-23T16:06:24.018 DEBUG:teuthology.orchestra.run.smithi114:> test -s /home/ubuntu/cephtest/cephadm && test $(stat -c%s /home/ubuntu/cephtest/cephadm) -gt 1000 && chmod +x /home/ubuntu/cephtest/cephadm
2023-11-23T16:06:24.036 DEBUG:teuthology.orchestra.run:got remote process result: 1
```

Replicated locally:

```
$ curl --silent https://raw.githubusercontent.com/ceph/ceph/reef/src/cephadm/cephadm
404: Not Found
```

For quincy it works.

I tracked the problem to 18fcfb022d19af1e69addf1729d4593272e4ac20 which got merged as a part of this PR.

There was a follow-up but the fix is imperfect: https://github.com/ceph/ceph/pull/48232.

Reply:
@rzarzynski the first commit from https://github.com/ceph/ceph/pull/53105 was intended to address pulling the cephadm binary from stable branches where it is compiled (currently only reef). I had the PR on hold becuase reef -> main upgrades currently don't work anyway because main is version 18.0.0 and reef is 18.2.0 so cephadm considers it a downgrade and blocks the upgrade. I know you were working on the v19 kickoff stuff. What I can probably do is separate out the first commit of that PR and add some additional work to have the reef-x upgrade tests make use of it. Then we could test that with the v19 kickoff work when it's ready.

Reply:
PR for what I mentioned in previous comment https://github.com/ceph/ceph/pull/54721